# no-else-if

* Prohibit `else if`.
  > `else if` ã‚’ç¦æ­¢ã—ã¾ã™ã€‚

## Rule Details

* Here's a summary of why `else if` is disallowed.
  > ä»¥ä¸‹ã¯ã€`else if` ãŒãªãœç¦æ­¢ã•ã‚Œã‚‹ã‹ã®ã¾ã¨ã‚ã§ã™ã€‚

### syntax sugar of nested if statements<br>å…¥ã‚Œå­ã«ãªã£ãŸ if æ–‡ã®ç³–è¡£æ§‹æ–‡

* `else if` is the sugar-coated syntax for nesting if statements; wherever `else if` is written, there will be as many nested conditional branches as there are repeated `else if`.
  > `else if` ã¯ã€if æ–‡ã‚’ãƒã‚¹ãƒˆã•ã›ã‚‹æ™‚ã®ç³–è¡£æ§‹æ–‡ã§ã™ã€‚else if ãŒæ›¸ã‹ã‚Œã¦ã„ã‚‹å ´æ‰€ã«ã¯ã€else if ã‚’ç¹°ã‚Šè¿”ã—ãŸæ•°ã ã‘æ¡ä»¶åˆ†å²ã®ãƒã‚¹ãƒˆãŒå­˜åœ¨ã—ã¾ã™ã€‚

### complex nested structures waste the reader's brain resources<br>è¤‡é›‘ãªå…¥ã‚Œå­æ§‹é€ ã¯ã€èª­ã¿æ‰‹ã®è„³ã®ãƒªã‚½ãƒ¼ã‚¹ã‚’æµªè²»ã•ã›ã‚‹

* Complex nesting of if statements bloats the code reader's effort in chasing down conditional branches. If all ifs were not nested, there would not be a single stack of if statements in the code reader's brain, thus not wasting brain resources.
  > if æ–‡ã‚’è¤‡é›‘ã«ãƒã‚¹ãƒˆã•ã›ã‚‹ã®ã¯ã€ã‚³ãƒ¼ãƒ‰ã®èª­ã¿æ‰‹ãŒæ¡ä»¶åˆ†å²ã‚’è¿½ã„ã‹ã‘ã‚‹åŠ´åŠ›ã‚’è‚¥å¤§åŒ–ã•ã›ã¾ã™ã€‚ã‚‚ã—ã‚‚å…¨ã¦ã® if ãŒãƒã‚¹ãƒˆã—ãªã„ã®ã§ã‚ã‚Œã°ã€ã‚³ãƒ¼ãƒ‰ã‚’èª­ã‚€äººã®è„³å†…ã«ã¯ if æ–‡ã®ã‚¹ã‚¿ãƒƒã‚¯ã¯ã²ã¨ã¤ã‚‚ç©ã¾ã‚Œã‚‹ã“ã¨ã¯ãªã„ã®ã§ã€è„³ã®ãƒªã‚½ãƒ¼ã‚¹ã‚’ç„¡é§„ã«æµªè²»ã•ã›ãªã„ã§ã—ã‚‡ã†ã€‚

* In object-oriented programming, conditional branches with if statements can be designed without any nesting. We have decided to prohibit `else if` so that all engineers can work on programming without avoiding precise class design.
  > ã‚ªãƒ–ã‚¸ã‚§ã‚¯ãƒˆæŒ‡å‘ãƒ—ãƒ­ã‚°ãƒ©ãƒŸãƒ³ã‚°ã§ã¯ã€if æ–‡ã«ã‚ˆã‚‹æ¡ä»¶åˆ†å²ã‚’ä¸€åˆ‡ãƒã‚¹ãƒˆã•ã›ã‚‹äº‹ç„¡ãè¨­è¨ˆã§ãã¾ã™ã€‚å…¨ã‚¨ãƒ³ã‚¸ãƒ‹ã‚¢ãŒç²¾å¯†ãªã‚¯ãƒ©ã‚¹è¨­è¨ˆã‚’å›é¿ã™ã‚‹ã“ã¨ãªããƒ—ãƒ­ã‚°ãƒ©ãƒŸãƒ³ã‚°ä½œæ¥­ã«å–ã‚Šçµ„ã‚ã‚‹æ§˜ã«ã™ã‚‹ç‚ºã«ã€`else if` ã‚’ç¦æ­¢ã™ã‚‹äº‹ã«ã—ã¾ã—ãŸã€‚


## How to Refactor to Purge `else if`

* To use `early returns` and `guard clauses` can avoid `else if` in many cases.
  > æ—©æœŸãƒªã‚¿ãƒ¼ãƒ³ã‚„ã‚¬ãƒ¼ãƒ‰ç¯€ã‚’åˆ©ç”¨ã™ã‚Œã°ã€å¤šãã®ã‚±ãƒ¼ã‚¹ã§ `else if` ã‚’å›é¿ã§ãã¾ã™ã€‚

ğŸ˜© Examples of **incorrect** code for `no-else-if` rule:

```js
function method (firstCondition, secondCondition) {
  if (firstCondition) {
    return createReturnValueAs1stCondition()
  } else if (secondCondition) {
    return createReturnValueAs2ndCondition()
  } else {
    return null
  }
}
```

ğŸ˜† Examples of **correct** code for `no-else-if` rule:

```js
function method (firstCondition, secondCondition) {
  if (firstCondition) {
    return createReturnValueAs1stCondition()
  }

  if (secondCondition) {
    return createReturnValueAs2ndCondition()
  }

  return null
}
```

* When multiple branches are to be made based on the value of a single variable, it is better to use object hashing.
  > ã²ã¨ã¤ã®å¤‰æ•°ã®å€¤ã«ã‚ˆã‚Šè¤‡æ•°ã®åˆ†å²ã‚’ã™ã‚‹å ´åˆã¯ã€ã‚ªãƒ–ã‚¸ã‚§ã‚¯ãƒˆãƒãƒƒã‚·ãƒ¥ã‚’ä½¿ã†æ–¹ãŒå„ªã‚Œã¦ã„ã¾ã™ã€‚

ğŸ˜© Examples of **incorrect** code for `no-else-if` rule:

```js
function method (condition) {
  if (condition === 'aaa') {
    return 1000
  } else if (condition === 'bbb') {
    return 2000
  } else if (condition === 'ccc') {
    return 3000
  } else {
    return 0
  }
}
```

ğŸ˜† Examples of **correct** code for `no-else-if` rule:

```js
function method (condition) {
  return objectHash[condition] ?? 0
}

const objectHash = {
  aaa: 1000,
  bbb: 2000,
  ccc: 3000,
}
```

## Version

* Since `0.0.0`
